### ISSUE TITLE
Groupby Shift produces incorrect result

### ISSUE THREAD

**Issue by kbushman** [dask/dask#8749]
**What happened**:

Data shifts to incorrect location when using groupby shift operation.

**What you expected to happen**:

Data should be shifted along the sorted index within each group.

**Minimal Complete Verifiable Example**:

```python
df1 = pd.DataFrame(
    {
        "a": [1, 2, 3, 4],
        "b": [2, 1, 2, 1],
        "c": [None, None, 10, 20]
    }
)
ddf1 = dd.from_pandas(df1, npartitions=1)
df2 = pd.DataFrame(
    {
        "a": [5, 6],
        "b": [2, 1],
        "c": [30, 40]
    }
)
ddf2 = dd.from_pandas(df2, npartitions=1)
df3 = pd.concat([df1, df2])
ddf3 = dd.concat([ddf1, ddf2])

df3 = df3.set_index("a").sort_index()
ddf3 = ddf3.set_index("a")


with pytest.warns(UserWarning):
    assert_eq(
        df3.groupby("b")["c"].shift(1),
        ddf3.groupby("b")["c"].shift(1),
    )
```

**Anything else we need to know?**:

There is a comment in the source code that "If the grouper does not align with the index then this causes a full shuffle.  The order of rows within each group may not be preserved." Although this is fine for apply/transform, preserving order is vital to the correctness of the shift operation.

**Potential solution**:

Sort each partition after shuffling. From dask/dataframe/groupby.py shift method:

```python
        if should_shuffle:
            sort_idx = df.index.name
            df2, by = self._shuffle(meta)
            df2 = df2.map_partitions(lambda x: x.sort_values(sort_idx))
```

**Environment**:

- Dask version: 2022.2.0
- Python version: 3.8.2
- Operating System: Mac OS
- Install method (conda, pip, source): pip

**Comment by beckernick** [comment]
I ran into this problem while working on #8659 in https://github.com/dask/dask/pull/8659#issuecomment-1037266953

I'm not sure sorting each partition after a full shuffle would resolve this problem, as the partitions themselves may be different. A naive approach could be to create a monotonically increasing global index column and sort based on that.

```python
import dask.dataframe as dd
import pandas as pd
​
df = pd.DataFrame({
    "a":[0,0,0,1,1,2,1,0,2,1,0],
    "b":[10,3,3,5,5,4,2,9,3,10,4]
})
ddf = dd.from_pandas(df, 2) #.reset_index(drop=True)
​
# pandas
df["lagged"] = df.groupby("a").rank()
​
# Dask
ddf["monotonic"] = 1
ddf["monotonic"] = ddf["monotonic"].cumsum()
​
def func(df):
    ranked = df.groupby("a")[["b"]].rank()
    df["lagged"] = ranked
    return df
​
​
ddf = ddf.shuffle(on="a").map_partitions(func).sort_values(["monotonic"])
ddf = ddf.drop("monotonic", axis=1)
​
pd.testing.assert_frame_equal(
    ddf.compute(),
    df
)
​
print(ddf.compute())
print(df)
    a   b  lagged
0   0  10     5.0
1   0   3     1.5
2   0   3     1.5
3   1   5     2.5
4   1   5     2.5
5   2   4     2.0
6   1   2     1.0
7   0   9     4.0
8   2   3     1.0
9   1  10     4.0
10  0   4     3.0
    a   b  lagged
0   0  10     5.0
1   0   3     1.5
2   0   3     1.5
3   1   5     2.5
4   1   5     2.5
5   2   4     2.0
6   1   2     1.0
7   0   9     4.0
8   2   3     1.0
9   1  10     4.0
10  0   4     3.0
```

cc @jsignell , interested in your thoughts. Perhaps further discussion should move to this issue rather than the groupby.rank PR

**Comment by jsignell** [comment]
Hmmm for the original issue I am not totally convinced that the shift is doing the wrong thing. The `c` value for each index value matches that in pandas. It is just that the index is not in order. 

```python
expected = df3.groupby("b")["c"].shift(1)
actual = ddf3.groupby("b")["c"].shift(1).compute()

In [5]: actual
Out[5]: 
a
2     NaN
4     NaN
6    20.0
1     NaN
3     NaN
5    10.0
Name: c, dtype: float64

In [6]: expected
Out[6]: 
a
1     NaN
2     NaN
3     NaN
4     NaN
5    10.0
6    20.0
Name: c, dtype: float64

**Comment by kbushman** [comment]
Hmm, interesting. I just reran it a few times and it seems to be non-deterministic. Sometimes I get the same result that you did (which I agree is the expected result). But other times I get the following result:

```python
a
1     NaN
2     NaN
3     NaN
4     NaN
5    10.0
6    20.0
Name: c, dtype: float64

a
6     NaN
2    40.0
4     NaN
5     NaN
1    30.0
3     NaN
Name: c, dtype: float64
```

**Comment by jsignell** [comment]
Ah ok. I admit it was a little bit wishful thinking. Thanks for indulging me.

**Comment by kori73** [comment]
I have opened a PR with the within-partition-sorting approach. I can also offer some help if you would like to go with the global index (or some other) approch to fix the issue discussed on #8659.

**Comment by jsignell** [comment]
Thanks @kori73! So this is closed by #8782
